### PR TITLE
Add PDF custom fonts

### PR DESCRIPTION
**Pull request by tsiegleauq**
Adds virtual file system with custom fonts to the pdfmake routine.
Uploading own fonts and use them in the PDF is now possible

also: Ships used fira-sans and fira-sans-condensed fonts as assets.

### CODE REVIEW

File: client/src/app/core/services/http.service.ts [hunk 1]
```diff
@@ -60,21 +60,26 @@ export class HttpService {
         method: HTTPMethod,
         data?: any,
         queryParams?: QueryParams,
-        customHeader?: HttpHeaders
+        customHeader?: HttpHeaders,
+        responseType?: string
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/http.service.ts]: What is allowed to provide here? Does somewhere exist a list of content types? For now, you may lock this to all needed types. Maybe define a `type ContentType = 'json' | '...' | ...;` on top of the file.

File: client/src/app/core/services/http.service.ts [hunk 2]
```diff
@@ -60,21 +60,26 @@ export class HttpService {
...
+        if (!responseType) {
+            responseType = 'json';
         }
 
         const url = path + formatQueryParams(queryParams);
         const options = {
             body: data,
-            headers: customHeader ? customHeader : this.defaultHeaders
+            headers: customHeader ? customHeader : this.defaultHeaders,
+            responseType: responseType as 'json'
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/http.service.ts]: Nice ;)

File: client/src/app/core/services/pdf-document.service.ts [hunk 3]
```diff
@@ -31,10 +32,84 @@ export class PdfDocumentService {
...
+
+    /**
+     * Define the pdfmake virtual file system for fonts
+     *
+     * @returns the vfs-object
+     */
+    private async getVfs(): Promise<object> {
+        const availableFonts = this.configService.instant('fonts_available');
+
+        const fontPathList: string[] = Array.from(
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/pdf-document.service.ts]: Why is this necessary? You may give the type of the config object: `const availableFonts = this.configService.instant<string[]>('fonts_available');`, then no casting must be done.

File: client/src/app/core/services/http.service.ts (hunk 2 above)
Comment (tsiegleauq) [reply client/src/app/core/services/http.service.ts]: https://github.com/angular/angular/issues/18586

File: client/src/app/core/services/http.service.ts (hunk 1 above)
Comment (tsiegleauq) [reply client/src/app/core/services/http.service.ts]: string is the correct type, there might be a lot of response types.
I set it in the documentation, sry about that.
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType

File: client/src/app/core/services/pdf-document.service.ts (hunk 3 above)
Comment (tsiegleauq) [reply client/src/app/core/services/pdf-document.service.ts]: instant does not yet support templates. Currect syntax would be `const availableFonts = this.configService.instant('fonts_available') as string[]` or `const availableFonts: string[] = this.configService.instant('fonts_available');`

However:
new Set(myList) returns a "Set"-object without redundancies, which needs to be converted to an array.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

File: client/src/app/core/services/pdf-document.service.ts [hunk 4]
```diff
@@ -31,10 +32,85 @@ export class PdfDocumentService {
...
+            // new Set will create an object that needs to be converted into an array
+            new Set(
+                availableFonts
+                    .map(available => this.configService.instant(available))
+                    .map(font => font.path || `/${font.default}`)
+            )
+        );
+
+        const promises = fontPathList.map(fontPath => {
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/core/services/pdf-document.service.ts]: You need to click a button to convert something to pdf, so the config should always be faster than any pdf generation

File: client/src/app/core/services/pdf-document.service.ts [hunk 5]
```diff
@@ -1,11 +1,12 @@
+import { HttpHeaders } from '@angular/common/http';
 import { Injectable } from '@angular/core';
 import { saveAs } from 'file-saver';
 
 import * as pdfMake from 'pdfmake/build/pdfmake';
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/core/services/pdf-document.service.ts]: I tried it out and apparently
`import  pdfMake  from 'pdfmake/build/pdfmake';`
also works. But
`import { pdfMake } from 'pdfmake/build/pdfmake';`
does not. So, it's still ugly, but prettier.

File: client/src/app/core/services/pdf-document.service.ts [hunk 6]
```diff
@@ -31,10 +32,85 @@ export class PdfDocumentService {
...
+    }
+
+    /**
+     * Converts a given blob to base64
+     *
+     * @param file File as blob
+     * @returns a promise to the base64 as string
+     */
+    private async convertUrlToBase64(url: string): Promise<string> {
+        const headers = new HttpHeaders().set('Content-Type', 'application/font-woff');
```
[the remark this answers is not in the document]
Comment (tsiegleauq) [reply client/src/app/core/services/pdf-document.service.ts]: I changed to use empty http headers. They are just to `get` a file from the server.

**Comment by tsiegleauq** [comment]
Cannot reproduce. Perhaps it was because of differences to the current master?
Made the required changes and rebased.

File: client/src/app/core/services/pdf-document.service.ts (hunk 3 above)
Comment (FinnStutzenstein) [reply client/src/app/core/services/pdf-document.service.ts]: Ok, can you add the generic parameter to instant? Just like get, then the interface in the config service is consistant..

File: client/src/assets/styles/font-variables.scss [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+/** Condensed Font */
+$font-condensed: 'OSFont Condensed';
+
+$font-condensed-regular: url('assets/fonts/fira-sans-condensed-latin-400.woff') format('woff');
+$font-weight-condensed-regular: 400;
+
+$font-condensed-italic: url('assets/fonts/fira-sans-condensed-latin-400italic.woff') format('woff');
+$font-weight-condensed-italic: 400;
+
+$font-condensed-medium: url('assets/fonts/fira-sans-condensed-latin-500.woff') format('woff');
```
Comment (emanuelschuetze) [inline client/src/assets/styles/font-variables.scss]: As condensed font we need one file only: fira-sans-condensed-latin-400.woff. This file we will add later as 5th font config value (for title font). You can remove the other condensed files.
